crates/red_knot_python_semantic/src/types.rs

```diff
@@ -471,6 +471,137 @@ impl<'db> Type<'db> {
         self == other
     }
 
+    /// Return true if this type and `other` have no common elements.
+    ///
+    /// Note: This function aims to have no false positives, but might return
+    /// wrong `false` answers in some cases.
+    pub(crate) fn is_disjoint_from(self, db: &'db dyn Db, other: Type<'db>) -> bool {
+        match (self, other) {
+            (Type::Never, _) | (_, Type::Never) => true,
+
+            (Type::Any, _) | (_, Type::Any) => false,
+            (Type::Unknown, _) | (_, Type::Unknown) => false,
+            (Type::Unbound, _) | (_, Type::Unbound) => false,
+            (Type::Todo, _) | (_, Type::Todo) => false,
+
+            (ty @ (Type::Function(..) | Type::Module(..) | Type::Class(..)), other)
+            | (other, ty @ (Type::Function(..) | Type::Module(..) | Type::Class(..))) => {
+                ty != other
+            }
```

Comment:
It just occurred to me now on second read-through that this case is incorrect if it comes before the union case, because the union of two `Type::Function` is not disjoint from one of those same `Type::Function`.

I think this case is really the same as the `None` and `Literal` types case below, and could probably just be merged with it?

Would be good to add a test case that would have caught this (to create a union of function or class types, define them in if/else branches.)

Reply:
> It just occurred to me now on second read-through that this case is incorrect if it comes before the union case, because the union of two `Type::Function` is not disjoint from one of those same `Type::Function`.

Ah, you're right.

> I think this case is really the same as the `None` and `Literal` types case below, and could probably just be merged with it?

:+1: 

> Would be good to add a test case that would have caught this (to create a union of function or class types, define them in if/else branches.)

Yes. I added a test that failed with the previous version.